src/plugins/intel_cpu/src/nodes/scaled_attn.cpp

```diff
@@ -806,6 +824,288 @@ bool ScaledDotProductAttention::isSupportedOperation(const std::shared_ptr<const
     return true;
 }
 
+void ScaledDotProductAttention::assignState(const std::shared_ptr<VariableStateKVcache>& state, int idx) {
+    auto inputNumber = getOriginalInputsNumber();
+    if (inputNumber - 2 == static_cast<size_t>(idx)) {
+        m_k_state = state;
+    } else if (inputNumber - 1 == static_cast<size_t>(idx)) {
+        m_v_state = state;
+    } else {
+        OPENVINO_THROW(
+            "Unexpected idx ", idx , " for a state in a node with type: ", getTypeStr(), " and name ", getName());
+    }
+}
+
+void ScaledDotProductAttention::gatherConcatPastkv(const MemoryPtr& mem_cur_k, const MemoryPtr& mem_cur_v, const MemoryPtr& mem_beam_idx) {
+    PlainTensor cur_k;
+    cur_k.reset(mem_cur_k);
+    if (!m_config.config.permute_axes.empty())
+        cur_k = cur_k.permute(m_config.config.permute_axes);
+
+    updateBeamTable(mem_beam_idx, cur_k.size(2));
+    updatePastkv(mem_cur_k, mem_cur_v);
+}
+
+void ScaledDotProductAttention::updateBeamTable(const MemoryPtr& mem_beam_idx, size_t L1) {
+    std::vector<size_t> order = {0, 1, 2, 3};
+    if (!m_config.config.permute_axes.empty()) {
+        order = m_config.config.permute_axes;
+    }
+    PlainTensor beam_idx, beam_table_k, beam_table_v;
+    auto hidden_state_k = m_k_state->hidden_state_mem();
+    auto hidden_state_v = m_v_state->hidden_state_mem();
+    beam_idx.reset(mem_beam_idx);
+
+    auto B = beam_idx.size(0);
+    size_t L0 = 0;
+    auto is_reset = m_k_state->is_reset_state() || m_v_state->is_reset_state();
+    if (is_reset) {
+        OPENVINO_ASSERT(m_k_state->is_reset_state() == m_v_state->is_reset_state(),
+            "KV state must be reset simultaneously, please also reset state for ",
+            (m_k_state->is_reset_state() ? "V" : "K"));
+        auto inputNumber = getOriginalInputsNumber();
+        auto&& init_graph_v_dims = getParentEdgeAt(inputNumber - 1)->getMemory().getStaticDims();
+        L0 = init_graph_v_dims.at(order[2]);
+        auto B_init = init_graph_v_dims.at(order[0]);
+        OPENVINO_ASSERT(B == B_init, "beam idx batch: ", B, " is not equal to batch of init subgraph: ", B_init);
+    } else if (hidden_state_k) {
+        auto block_desc = hidden_state_k->getDescWithType<BlockedMemoryDesc>();
+        auto&& dims = block_desc->getShape().getStaticDims();
+        auto B_state = dims[0];
+        L0 = dims[1];
+        OPENVINO_ASSERT(B == B_state, "beam idx batch: ", B, " is not equal to batch of state: ", B_state);
+    }
```

Comment:
MemoryInputSDPA node can always give correct shape for past_key/value, whether is_reset or not, so why not simply retrieve L0 from input edge?

Reply:
Updated.